ql/src/java/org/apache/hadoop/hive/ql/exec/vector/mapjoin/VectorMapJoinAntiJoinMultiKeyOperator.java

```diff
@@ -0,0 +1,400 @@
+          joinResult = JoinUtil.JoinResult.MATCH;
+        } else {
+          byte[] keyBytes = currentKeyOutput.getData();
+          int keyLength = currentKeyOutput.getLength();
+          // LOG.debug(CLASS_NAME + " processOp all " + displayBytes(keyBytes, 0, keyLength));
+          joinResult = hashSet.contains(keyBytes, 0, keyLength, hashSetResults[0]);
+          // reverse the join result from hash table for anti join.
+          if (joinResult == JoinUtil.JoinResult.NOMATCH) {
+            joinResult = JoinUtil.JoinResult.MATCH;
+          } else if (joinResult == JoinUtil.JoinResult.MATCH) {
+            joinResult = JoinUtil.JoinResult.NOMATCH;
+          }
+        }
+
+        /*
+         * Common repeated join result processing.
+         */
+
+        if (LOG.isDebugEnabled()) {
+          LOG.debug(CLASS_NAME + " batch #" + batchCounter + " repeated joinResult " + joinResult.name());
+        }
+        finishAntiRepeated(batch, joinResult, hashSetResults[0]);
+      } else {
+
+        /*
+         * NOT Repeating.
+         */
+
+        if (LOG.isDebugEnabled()) {
+          LOG.debug(CLASS_NAME + " batch #" + batchCounter + " non-repeated");
+        }
+
+        // We remember any matching rows in matches / matchSize.  At the end of the loop,
+        // selected / batch.size will represent both matching and non-matching rows for outer join.
+        // Only deferred rows will have been removed from selected.
+        int selected[] = batch.selected;
+        boolean selectedInUse = batch.selectedInUse;
+
+        int hashSetResultCount = 0;
+        int allMatchCount = 0;
+        int spillCount = 0;
+
+        /*
+         * Multi-Key specific variables.
+         */
+
+        Output temp;
+
+        // We optimize performance by only looking up the first key in a series of equal keys.
+        boolean haveSaveKey = false;
+        JoinUtil.JoinResult saveJoinResult = JoinUtil.JoinResult.NOMATCH;
+
+        // Logical loop over the rows in the batch since the batch may have selected in use.
+        for (int logical = 0; logical < inputLogicalSize; logical++) {
+          int batchIndex = (selectedInUse ? selected[logical] : logical);
+
+          /*
+           * Multi-Key get key.
+           */
+
+          // Generate binary sortable key for current row in vectorized row batch.
+          keyVectorSerializeWrite.setOutput(currentKeyOutput);
+          keyVectorSerializeWrite.serializeWrite(batch, batchIndex);
+          boolean isAnyNull = keyVectorSerializeWrite.getHasAnyNulls();
+
+          // LOG.debug(CLASS_NAME + " currentKey " +
+          //      VectorizedBatchUtil.displayBytes(currentKeyOutput.getData(), 0, currentKeyOutput.getLength()));
+
+          /*
+           * Equal key series checking.
+           */
+
+          if (isAnyNull || !haveSaveKey || !saveKeyOutput.arraysEquals(currentKeyOutput)) {
+
+            // New key.
+
+            if (haveSaveKey) {
```

Comment:
simplification?

Reply:
I have kept the code same as Semi join. may be we can take up this refactoring as part of merging the code.